### ISSUE TITLE
Incorrect mappings in GNU standard library, LLVM standard library and Boost

### ISSUE THREAD

**Issue by correaa** [include-what-you-use/include-what-you-use#1561]
In a project of mine (https://github.com/correaa/boost-multi/blob/master/.iwyu-test.imp) I detected a bunch of extra false positives, 

in GNU standard library:

```lml
    { "include": ["@<bits/chrono.h>", "private", "<chrono>", "public"] },
    { "include": ["@<bits/utility.h>", "private", "<tuple>", "public"] },
    { "include": ["@<bits/memory_resource.h>", "private", "<memory_resource>", "public"] },
    { "include": ["@<assert.h>", "public", "<cassert>", "public"] },
    { "include": ["@<stdint.h>", "public", "<cstdint>", "public"] },
    { "include": ["@<math.h>", "public", "<cmath>", "public"] },
    { "include": ["@<ext/alloc_traits.h>", "private", "<memory>", "public"]},
    { "include": ["@<shared_mutex>", "public", "<utility>", "public"]},
    { "include": ["@<__assert>", "private", "<cassert>", "public"] },
```

in LLVM standard library:

```yml
    { "include": ["@<__assert>", "private", "<cassert>", "public"] },
    { "include": ["@<__fwd/ostream.h>", "private", "<ostream>", "public"] },
    { "include": ["@<__fwd/sstream.h>", "private", "<sstream>", "public"] }
```

and in Boost (Boost.Test)

```yml
    { "include": ["@<boost/preprocessor/arithmetic/dec.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/arithmetic/limits/dec_256.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/seq/elem.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/seq/size.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/variadic/elem.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/control/expr_iif.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/detail/auto_rec.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/detail/limits/auto_rec_256.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/repetition/detail/for.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/repetition/detail/limits/for_256.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/repetition/for.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/seq/limits/elem_256.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/seq/limits/size_256.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/tuple/elem.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/variadic/limits/elem_64.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/comparison/not_equal.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/control/iif.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/logical/bool.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/logical/compl.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/comparison/limits/not_equal_256.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/preprocessor/logical/limits/bool_256.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tools/assertion.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tools/interface.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tools/fpc_op.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tools/old/interface.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tree/auto_registration.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tree/decorator.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tree/test_unit.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/tree/global_fixture.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/unit_test_suite.hpp>", "private", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/framework.hpp>", "public", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/unit_test_log.hpp>", "public", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/utils/basic_cstring/basic_cstring.hpp>", "public", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/utils/basic_cstring/basic_cstring_fwd.hpp>", "public", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/utils/lazy_ostream.hpp>", "public", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/test/utils/wrap_stringstream.hpp>", "public", "<boost/test/unit_test.hpp>", "public"] },
    { "include": ["@<boost/mpl/aux_/preprocessed/gcc/list.hpp>", "private", "<boost/mpl/list.hpp>", "public"] },
```

and in Boost MultiArray (questionable), since MultiArray uses iterator_facade (from Boost.Iterator) effectively as an implementation detail

```yml
    { "include": ["@<boost/iterator/iterator_facade.hpp>", "private", "<boost/multi_array.hpp>", "public"] },
```

**Which of these are worth to open a PR?**

**Comment by kimgr** [comment]
Most of these look like they make sense in the context of your project, but maybe not generally? Mapping e.g. `boost/preprocessor/logical/limits/bool_256.hpp` to `boost/test/unit_test.hpp` doesn't seem like it will be useful to the general public.

In general, I would prefer to update mappings using repeatable scripts rather than hand-patching. There's a bunch of shell pipelines in `iwyu_include_picker.cc` and the `*.imp` files, that I'm hoping to port to Python, so they can express any manual overrides more explicitly.

**Comment by correaa** [comment]
> Most of these look like they make sense in the context of your project, but maybe not generally? Mapping e.g. `boost/preprocessor/logical/limits/bool_256.hpp` to `boost/test/unit_test.hpp` doesn't seem like it will be useful to the general public.
> 

My criterion to add things is that what iwyu reports is not something that is explicitly used in the cpp file. 
But I understand, Boost Test is mostly a library of macros and who know what these macros expand to. 
It is general in the sense that anyone making even minimal use of Boost Test will see all these extra headers needed. 

What about the other (non-Boost) sets?

> 
> 
> In general, I would prefer to update mappings using repeatable scripts rather than hand-patching. There's a bunch of shell pipelines in `iwyu_include_picker.cc` and the `*.imp` files, that I'm hoping to port to Python, so they can express any manual overrides more explicitly.

ok, sure. I don't know how to run such scripts but I guess you will need actual code based to generate these. 

In that case you can use my project  https://github.com/correaa/boost-multi that can use iwyu through Cmake. 

https://github.com/correaa/boost-multi/blob/eb7da4e3d3273f417377db407a6c366183ab0f5b/.gitlab-ci.yml#L394

https://github.com/correaa/boost-multi/blob/master/.iwyu-test.imp

Cheers,
A

**Comment by kimgr** [comment]
> ok, sure. I don't know how to run such scripts but I guess you will need actual code based to generate these.

See the `mapgen/` directory for a few examples of what I had in mind: https://github.com/include-what-you-use/include-what-you-use/tree/master/mapgen.

They build mappings based on simple pattern matching over library headers, and knowledge about the individual library conventions.

**Comment by firewave** [comment]
The Boost part is essentially a duplicate of #1553, #1046 and #227. I will elaborate on it in the latter.

**Comment by correaa** [comment]
> > ok, sure. I don't know how to run such scripts but I guess you will need actual code based to generate these.
> 
> See the `mapgen/` directory for a few examples of what I had in mind: https://github.com/include-what-you-use/include-what-you-use/tree/master/mapgen.
> 
> They build mappings based on simple pattern matching over library headers and knowledge about the individual library conventions.

It sounds like a tool for experts. Also, this will produce a  `.imp` file, right?

It is difficult to know the conventions of each library; it seems as difficult as enumerating the rules manually. 

(Ideally, each library should declare its own conventions using IWYU as the defacto tool for automatic inclusion. (like I am doing with my library, e.g. https://github.com/correaa/boost-multi/blob/master/include/boost/multi/detail/layout.hpp#L8)

Also, what about the GNU and LLVM library rules I suggested in the first post?

**Comment by firewave** [comment]
> Ideally, each library should declare its own conventions using IWYU as the defacto tool for automatic inclusion.

Yes, that would be the best case scenario. Qt recently introduced IWYU annotations. That might not solve all the issues though. See #94 for more details.

**Comment by kimgr** [comment]
> (Ideally, each library should declare its own conventions using IWYU as the defacto tool for automatic inclusion.
> (like I am doing with my library, e.g. https://github.com/correaa/boost-multi/blob/master/include/boost/multi/detail/layout.hpp#L8)

Yeah, this is super cool. I still think of IWYU as a hobby project, so I'm always a bit surprised to see it used as a defacto tool.

There's a loose plan in IWYU (#1271) to dynamically pick up mapping files from the -I include path. The idea is that libraries should be able to ship with official mappings that IWYU will discover and apply as it traverses the include path. As laid out in that issue, there are a bunch of small blockers, but maybe this summer vacation I'll make some headway, who knows.

> Also, what about the GNU and LLVM library rules I suggested in the first post?

I guess I'll need to think more about those. I've tried to reject them on the basis that they should be generated in the first place. Now I'm spending some time hacking on an actual repeatable generator for libstdc++ so we don't have to mess around with ad-hoc patches for invidual headers. Hopefully I'll have something soon.

**Comment by correaa** [comment]
> > (Ideally, each library should declare its own conventions using IWYU as the defacto tool for automatic inclusion.
> > (like I am doing with my library, e.g. https://github.com/correaa/boost-multi/blob/master/include/boost/multi/detail/layout.hpp#L8)
> 
> Yeah, this is super cool. I still think of IWYU as a hobby project, so I'm always a bit surprised to see it used as a defacto tool.

Is there something else that remotely tries to solve the problems of headers?
This seems to be an unsolved problem in C++ (and C, particularly bad for C++ because of its complexity).
Most programs can get away with poorly maintained includes because people (including me) would add headers until it works.

One of the reasons I used it is because it integrates very nicely with CMake.
Although I have found problems reporting errors back to cmake to stop builds.
Also I don't understand why the tool doesn't recursively check `hpp` files.

> 
> There's a loose plan in IWYU (#1271) to dynamically pick up mapping files from the -I include path. The idea is that libraries should be able to ship with official mappings that IWYU will discover and apply as it traverses the include path. As laid out in that issue, there are a bunch of small blockers, but maybe this summer vacation I'll make some headway, who knows.

Yes, that would be nice for libraries that don't want to modify the source code.
I guess some kind of default named file `.iwyu.imp` or with same as the header `mylib.imp` or `mylib.iwyu` automatically associated with `mylib.hpp`.

> 
> > Also, what about the GNU and LLVM library rules I suggested in the first post?
> 
> I guess I'll need to think more about those. I've tried to reject them on the basis that they should be generated in the first place. Now I'm spending some time hacking on an actual repeatable generator for libstdc++ so we don't have to mess around with ad-hoc patches for invidual headers. Hopefully I'll have something soon.

Ah, ok, same problem.

**Comment by kimgr** [comment]
I'm calling this fixed for the GNU libstdc++ as of #1606 and #1694.

The LLVM libc++ is tracked in #1271 and associated issues.

That leaves the Boost headers. I'll keep this issue open as a feature request to add a Boost mapping generator.

**Comment by firewave** [comment]
Boost is covered by #227, so this can be closed.

**Comment by kimgr** [comment]
Alright, thanks. Closing.
